File: frontend/src/api/analysis.ts
import { request } from './client';

export interface ColumnGroup { data_type_key?: string; table: string; label: string; row_count?: number; duration_sec?: number; columns: ColumnItem[]; }
export interface ColumnItem { key: string; label: string; unit: string; scale_factor: number; }
export interface AlignedData { times: string[]; ref_secs: number[]; series: Record<string, { label: string; unit: string; scale_factor: number; is_numeric: boolean; table: string; values: (number | null)[]; text_values?: (string | null)[] }>; mask?: boolean[]; segments?: { start: number; end: number }[]; }
export interface FilterCondition { column: string; op: 'gt' | 'gte' | 'lt' | 'lte' | 'eq' | 'between'; value: number | null; min_val: number | null; max_val: number | null; }
export interface FilterSpec { logic: 'and' | 'or'; conditions: FilterCondition[]; }
export interface FilterPreset { id: number; model_id: number; name: string; config: FilterSpec; }
export interface FlightStats { duration_sec: number; start_time: string; end_time: string; drone_id: string; name: string; }
export interface Preset { id: number; model_id: number; name: string; columns: string[]; }
export interface CorrelationData { columns: string[]; labels: string[]; matrix: number[][]; }
export interface AnomalyData { times: string[]; values: number[]; anomaly_indices: number[]; upper_bound: number[]; lower_bound: number[]; label: string; unit: string; }
export interface CompareSeries { flight_id: number; name: string; times_sec: number[]; values: number[]; label: string; unit: string; }
export const getRegistryColumns = (modelId: number) => request<{ columns: ColumnGroup[] }>(`/registry/columns?model_id=${modelId}`);
export const getColumns = (flightId: number) => request<{ columns: ColumnGroup[] }>(`/flights/${flightId}/columns`);
export const getAlignedData = (flightId: number, columnKeys: string[], filter?: FilterSpec) => request<AlignedData>(`/flights/${flightId}/aligned`, { method: 'POST', body: JSON.stringify({ column_keys: columnKeys, filter: filter || undefined }) });
export const getStats = (flightId: number) => request<FlightStats>(`/flights/${flightId}/stats`);
export const getCorrelation = (flightId: number, columnKeys: string[]) => request<CorrelationData>(`/flights/${flightId}/correlation`, { method: 'POST', body: JSON.stringify({ column_keys: columnKeys }) });
export const getAnomaly = (flightId: number, columnKey: string, windowSize = 30, sigma = 3.0) => request<AnomalyData>(`/flights/${flightId}/anomaly`, { method: 'POST', body: JSON.stringify({ column_key: columnKey, window_size: windowSize, sigma }) });
export const getCompare = (flightIds: number[], columnKey: string) => request<{ series: CompareSeries[] }>('/compare', { method: 'POST', body: JSON.stringify({ flight_ids: flightIds, column_key: columnKey }) });
export const listPresets = (modelId: number) => request<{ presets: Preset[] }>(`/presets?model_id=${modelId}`);
export const createPreset = (modelId: number, name: string, columns: string[]) => request<Preset>('/presets', { method: 'POST', body: JSON.stringify({ model_id: modelId, name, columns }) });
export const deletePreset = (id: number) => request('/presets/' + id, { method: 'DELETE' });
export const listFilterPresets = (modelId: number) => request<{ presets: FilterPreset[] }>(`/filter-presets?model_id=${modelId}`);
export const createFilterPreset = (modelId: number, name: string, config: FilterSpec) => request<FilterPreset>('/filter-presets', { method: 'POST', body: JSON.stringify({ model_id: modelId, name, config }) });
export const deleteFilterPreset = (id: number) => request('/filter-presets/' + id, { method: 'DELETE' });

File: backend/api/desktop/routers/analysis.py
"""Flight analysis, registry, and preset routes."""

from fastapi import APIRouter, HTTPException

from backend import analysis
from backend.api.desktop.schemas import (
    AlignedRequest,
    AnomalyRequest,
    CompareRequest,
    CorrelationRequest,
    FlightDataMatchesRequest,
    FilterPresetCreate,
    PresetCreate,
)
from backend.database import get_db
from backend.import_pipeline.format_configs import get_columns_for_model
from backend.repositories import presets as preset_repository


router = APIRouter()


@router.post("/api/flights/data-matches")
def get_data_matching_flights(req: FlightDataMatchesRequest):
    try:
        flight_ids = analysis.match_flights_by_data(
            req.model_id,
            req.flight_ids,
            req.filter,
        )
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    return {
        "flight_ids": flight_ids,
        "evaluated_count": len(set(req.flight_ids)),
        "matched_count": len(flight_ids),
    }


@router.get("/api/flights/{flight_id}/columns")
def get_columns(flight_id: int):
    return {"columns": analysis.get_columns_for_flight_api(flight_id)}


@router.post("/api/flights/{flight_id}/aligned")
def get_aligned(flight_id: int, req: AlignedRequest):
    return analysis.get_aligned_data(flight_id, req.column_keys, filter_spec=req.filter)


@router.get("/api/flights/{flight_id}/alerts")
def get_alerts(flight_id: int):
    return analysis.get_alerts(flight_id)


@router.get("/api/flights/{flight_id}/stats")
def get_stats(flight_id: int):
    return analysis.get_flight_stats(flight_id)


@router.post("/api/flights/{flight_id}/correlation")
def correlation(flight_id: int, req: CorrelationRequest):
    return analysis.get_correlation(flight_id, req.column_keys)


@router.post("/api/flights/{flight_id}/anomaly")
def anomaly(flight_id: int, req: AnomalyRequest):
    return analysis.get_anomalies(flight_id, req.column_key, req.window_size, req.sigma)


@router.post("/api/compare")
def compare(req: CompareRequest):
    return {"series": analysis.get_compare(req.flight_ids, req.column_key)}


@router.get("/api/registry/columns")
def registry_columns(model_id: int):
    conn = get_db()
    try:
        return {"columns": get_columns_for_model(conn, model_id)}
    finally:
        conn.close()


@router.get("/api/presets")
def list_presets(model_id: int):
    conn = get_db()
    try:
        return {"presets": preset_repository.list_column_presets(conn, model_id)}
    finally:
        conn.close()


@router.post("/api/presets")
def create_preset(req: PresetCreate):
    conn = get_db()
    try:
        result = preset_repository.save_column_preset(conn, req.model_id, req.name, req.columns)
        conn.commit()
        return result
    except Exception as exc:
        raise HTTPException(400, str(exc)) from exc
    finally:
        conn.close()


@router.delete("/api/presets/{preset_id}")
def delete_preset(preset_id: int):
    conn = get_db()
    try:
        preset_repository.delete_column_preset(conn, preset_id)
        conn.commit()
        return {"ok": True}
    finally:
        conn.close()


@router.get("/api/filter-presets")
def list_filter_presets(model_id: int):
    conn = get_db()
    try:
        return {"presets": preset_repository.list_filter_presets(conn, model_id)}
    finally:
        conn.close()


@router.post("/api/filter-presets")
def create_filter_preset(req: FilterPresetCreate):
    conn = get_db()
    try:
        result = preset_repository.save_filter_preset(conn, req.model_id, req.name, req.config)
        conn.commit()
        return result
    except Exception as exc:
        raise HTTPException(400, str(exc)) from exc
    finally:
        conn.close()


@router.delete("/api/filter-presets/{preset_id}")
def delete_filter_preset(preset_id: int):
    conn = get_db()
    try:
        preset_repository.delete_filter_preset(conn, preset_id)
        conn.commit()
        return {"ok": True}
    finally:
        conn.close()
